services/meminator-ruby/src/app.rb

```diff
@@ -11,66 +9,12 @@ class MeminatorApp < Sinatra::Application
 
   TRACER = OpenTelemetry.tracer_provider.tracer('MeminatorApp', '0.1.0')
 
-  IMAGE_MAX_WIDTH_PX = 1000
-  IMAGE_MAX_HEIGHT_PX = 1000
-  IMAGE_RESIZE_DIMENSIONS = "#{IMAGE_MAX_WIDTH_PX}x#{IMAGE_MAX_HEIGHT_PX}"
-
   post '/applyPhraseToPicture' do
-    request.body.rewind
-    work = JSON.parse(request.body.read)
-    logger.info "Received request: #{work}"
-
-    phrase = work.fetch("phrase") { return [400, "Missing 'phrase' key in JSON request body."] }
-    OpenTelemetry::Trace.current_span.set_attribute("app.phrase", phrase)
-
-    image_url = work.fetch("imageUrl") { return [400, "Missing 'imageUrl' key in JSON request body."] }
-    OpenTelemetry::Trace.current_span.set_attribute("app.imageUrl", image_url)
-
-    begin
-      image_url = URI.parse(image_url)
-      [URI::HTTP, URI::HTTPS].include?(image_url.class) or raise URI::Error.new("URL scheme must be HTTP or HTTPS.")
-    rescue URI::Error => e
-      OpenTelemetry::Trace.current_span.record_exception(e)
-      return [400, "Invalid 'imageUrl' value in JSON request body. #{e.message}"]
-    end
-
-    file_extension = File.extname(image_url.path)
-    downloaded_image = Tempfile.new(["downloaded_image", file_extension])
-    downloaded_image.binmode
-
-    TRACER.in_span('retrieve_image') do |span|
-      downloaded_image.write(URI.open(image_url).read)
-    end
-
-    meme_image = Tempfile.new(["meme_image", file_extension])
-
-    TRACER.in_span('convert') do |conversion_span|
-      memination_command = [
-        "convert", downloaded_image.path,
-        "-resize", IMAGE_RESIZE_DIMENSIONS,
-        "-gravity", "'North'",
-        "-pointsize", "48",
-        "-fill", "'white'",
-        "-undercolor", "'#00000080'",
-        "-font", "'Angkor-Regular'",
-        "-annotate", "0",
-        ('"' + phrase.upcase + '"'),
-        meme_image.path
-      ].join(" ")
-
-      conversion_span.set_attribute("app.subprocess.command", memination_command)
-
-      cmd_out, cmd_err, status = Open3.capture3(memination_command)
-      conversion_span.add_attributes({
-        "app.subprocess.returncode" => status.exitstatus,
-        "app.subprocess.stderr" => cmd_err,
-        "app.subprocess.stdout" => cmd_out,
-      })
+    phrase, image_url, errors = MemeGenerator.parse_request(request)
+    return [400, { errors: errors}.to_json] if !errors.empty?
```

Comment:
Are we not losing some of the custom instrumentation in this pattern?

Reply:
I did remove the custom attributes and the custom convert span in this PR.

I could add them back in, but there's a tickle in my memory now that it is the long download from S3 and entirely-opaque-to-auto-instrumentation shell out command that a student would be guided towards adding these custom instruments around.

I can't find the reference now, but in my head it is:

1. learn how to add auto-instrumentation (to backend-for-frontend)
2. learn how to add attributes to existing auto-instrumentation spans

Now that the student can see a single trace …
* Notice how meminator takes a long time? 
* What's up with that?
* Oh, there's a long call to an external service.
* Double-oh, there's this shell-out that I don't get a span for.

3. learn how to add custom spans around your own interesting business logic

Am I off the Rails† on this?

--- 

† See what I did there?

Reply:
@dmarchante and I chatted.

I'll add the `convert` span around the shell-out command for that to exist as the student's starting state. Scripts exist for other languages and those don't have the student create a custom span in the meminator service.